### DISCUSSION TITLE
AshPostgres' quirks

### DISCUSSION THREAD

**Discussion by almirsarajcic** [3Stones-io/omedis#287] [General]
These days we've had several chances to notice Ash trying to generate migrations for existing columns, indexes, or references.
I'm not sure how to solve this.

Today I wanted to fix the issue for some group_users index. My initial thought was to generate the migration and commit that, but the index already existed, so the command `mix ecto.migrate` crashed. So, I left the snapshot that was generated for group_users table and removed the newly generated Ecto migration. That seems to solve the issue.

Also, there is a `mix ash_postgres.squash_snapshots` command which probably should be executed every once in a while to make `mix ash_postgres.generate_migrations` run faster.

CC: @amos-kibet, @wintermeyer

Related: https://github.com/wintermeyer/omedis/pull/288 and https://github.com/wintermeyer/omedis/pull/289.

Let me know what you think about this.

**Comment by wintermeyer** [comment]
This looks like an Elixirforum problem/question. I have the feeling that the Ash gods have a solution for this.

**Reply by almirsarajcic** [reply]
You're right. And I just went there to post my question, but someone already did that.
https://elixirforum.com/t/ash-snapshot-migration-system/67137
It seems squashing is the way to go.
I've created a task for us (https://github.com/wintermeyer/omedis/issues/296) to add a check to make sure everything's up to date.

**Comment by amos-kibet** [comment]
Good point with squashing resource snapshots periodically.

**Comment by almirsarajcic** [comment]
Sometimes I like to run
```bash
mix ash_postgres.generate_migrations --dry-run
```
just to add some problems to my life when I get too comfortable.

If it shows some output, I do it without the `--dry-run` flag so it really generates the migration. Then, I run `mix ecto.migrate` and if it complains that table, index, or anything else already exists, I remove that migration as it's unnecessary.

The important step is the `mix ash_postgres.generate_migrations` command because it generates a resource snapshot that will prevent that migration from being generated again. Then I also run `mix ash_postgres.squash_snapshots` to remove any extra resource snapshots.

**Comment by almirsarajcic** [comment]
So many potential Elixir drops 😁
